lib/tracer/context.go

```diff
@@ -0,0 +1,204 @@
+package tracer
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+const (
+	// whether enable tracing
+	traceEnableParam = "trace_query"
+	// whether to print tracing with colors in stdout
+	traceEnableColorParam = "trace_query_color"
+	// specific trace ID. If omitted a random ID will be generated
+	traceIDParam = "trace_query_id"
+)
+
+// NewContext creates a new instance of the Context.
+// If enabled=false, all function calls to the returned
+// object will be no-op.
+func NewContext(r *http.Request) *Context {
+	defaultCtx := &Context{enabled: false}
+	if r == nil {
+		return defaultCtx
+	}
+
+	enabled := getBool(r, traceEnableParam)
+	id := r.FormValue(traceIDParam)
+
+	if !enabled && id == "" {
+		return defaultCtx
+	}
+
+	if id == "" {
+		id = newID()
+	}
+
+	enableColor := getBool(r, traceEnableColorParam)
+	now := time.Now()
+	return &Context{
+		id:      id,
+		color:   enableColor,
+		enabled: enabled,
+		updated: now,
+		start:   now,
+	}
+}
+
+var nextID = uint64(time.Now().UnixNano())
+
+func newID() string {
+	id := atomic.AddUint64(&nextID, 1)
+	return fmt.Sprintf("%08X", id)
+}
+
+// Context represents a tracing context.
+// Context must be created via NewContext func.
+// Each created context must be finalized via Done func.
+//
+// Context might contain sub-contexts (branches)
+// in order to build tree-like execution order.
+// To add a sub-context use Add func.
+type Context struct {
+	// id is an optional param and used only for the root context
+	id string
+	// msg is the message/title of the current context
+	msg string
+	// start is the time when Context was created
+	start time.Time
+	// enabled makes all function calls no-op if set to false
+	enabled bool
+	// color defines whether to print tracing with colors in stdout
+	color bool
+
```

Comment:
I think, that optional key=value pairs may be useful for context. 

It simplifies programmatic trace analyze.

 For instance, i want to group traces by top processed rows. Or by top executed function.

 Something like:
type KeyValue struct{
   Key string
   Value interface {}
}
```
func (c *Context) Done(msg string, kv KeyValue...){}...
...

func someQuery(){
  ctx.Done("processed parallel query", {"ts_len", 124124}, {"function", "up"})
}

```